### ISSUE TITLE
Lint idea: warn about implicit drop

### ISSUE THREAD

**Issue by m-mueller678** [rust-lang/rust-clippy#3915]
When writing low level code it is sometimes desirable to not have implicit drop calls.

For instance when interfacing with Lua one often has Rust function called by Lua, which in turn call Lua functions. One can execute a Lua function using [lua_call](https://www.lua.org/manual/5.3/manual.html#lua_call). Errors in the Lua code are handled using longjmp, which can jump to a handler in the original Lua caller thus skipping the Execution of the rest of the Rust function.
In such a situation one wants to make sure no code is executed implicitly after lua_call (because it might be skipped).
Ideally this lint can be enabled both for single values that should not be dropped implicitly as well as functions in which nothing should be dropped implicitly

**Comment by Manishearth** [comment]
Given that most people will not be using this lint, I am concerned about
the perf impact of the implementation, as currently all lint code is
unconditionally run (with lint _output_ being suppressed when lints are
disabled). This can and should be changed (
https://github.com/rust-lang/rust/issues/59024) but in the current state
lints requiring complex block analysis probably should not be added unless
they are on by default

**Comment by oli-obk** [comment]
Is that even safe? I'm thinking about a situation like below

```rust
fn foo(f: impl FnOnce()) {
    // do some setup
    catch_panic(f);
    // do some cleanup
}
```

and if you call this as

```rust
foo(|| {
    lua_call(...);
});
```

and `lua_call` does a longjmp, the cleanup from `foo` is never executed, even though execution continued. This isn't just about leaking memory and forgetting drops, this can cause actual UB.

**Comment by m-mueller678** [comment]
> Is that even safe? I'm thinking about a situation like below
> 
> ```rust
> fn foo(f: impl FnOnce()) {
>     // do some setup
>     catch_panic(f);
>     // do some cleanup
> }
> ```
> 
> and if you call this as
> 
> ```rust
> foo(|| {
>     lua_call(...);
> });
> ```
> 
> and `lua_call` does a longjmp, the cleanup from `foo` is never executed, even though execution continued. This isn't just about leaking memory and forgetting drops, this can cause actual UB.

lua_call is indeed unsafe. [Rust release notes 1.24.1](https://blog.rust-lang.org/2018/03/01/Rust-1.24.1.html) has some information on a related issue and seems to imply that usage is safe as long as:

- Lua never makes longjmp over Rust stackframes with non-copy types (I assume already dropped is fine)
- Rust code does not unwind into Lua code

rlua seems to handle this by only using [lua_pcall](https://www.lua.org/manual/5.3/manual.html#lua_pcall) which catches errors. I have not looked into it, but assume this messes with Lua error handlers. Error handlers in Lua are invoked immediately after an error is thrown e. g. to collect stacktraces. I think this can be worked around somewhat, but it is a niche feature anyways.
So this approach is an acceptable workaround for this usecase.

I still think this could be a useful feature once [rust-lang/rust#59024](https://github.com/rust-lang/rust/issues/59024#issue-418908450) is resolved.
